056-zk-kraft-migration.md

```diff
@@ -0,0 +1,247 @@
+
+### Finite State Machine (FSM)
+
+The migration process can be described with a Finite State Machine (FSM).
+The operator has a new `MigrationManager` component in charge of tracking the migration finite state machine.
+Transitioning across the states depends on the current state, stored in the `Kafka.status.migrationState` field, and the "input" represented by the user interaction with the operator through the following annotations:
+
+* `strimzi.io/kraft-migration`: to be applied on the `Kafka` custom resource with `true|false` values to start/end the overall migration process.
+* `strimzi.io/kraft-migration`: to be applied on the `KafkaNodePool`(s) custom resource with `dual|kraft|zk` values to represent the the desired "direction" in the migration process by the user:
+  * `dual`: allows to move the migration process forward to reach the point of the "dual-write" with KRaft controllers deployed but with metadata written to both ZooKeeper and KRaft. Brokers are also connected to both ZooKeeper nodes and KRaft controllers.
+  * `kraft`: allows to move the migration process forward from the "dual-write" to the end of the migration with ZooKeeper out of the picture.
+  * `zk`: allows to move the migration process back, when still possible, to have brokers working again with ZooKeeper ensemble only.
+
+The state machine is described in the following diagram:
+
+![Migration Finite State Machine](./images/056-zk-kraft-migration-fsm.png)
+
+The following sections explain the states and transitions in details.
+
+#### Phase 1: Provisioning the KRaft controller quorum
+
+Starting from a `NoMigration` (or null) state (as per `Kafka.status.migrationState` field), the user applies the `strimzi.io/kraft-migration: true` on the `Kafka` custom resource to make the operator aware of the migration "intention".
+The FSM moves to a `WaitControllersInMigration` state where the operator expects the user to create the KRaft controllers pool.
+The user creates the `KafkaNodePool` custom resource containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
+These controllers needed to be deployed in "migration mode", so the `KafkaNodePool` custom resource has to be annotate with `strimzi.io/kraft-migration: dual`.
+
+This way, the operator configures the controllers accordingly with the ZooKeeper migration flag and the connection to the ZooKeeper ensemble. 
+
+```shell
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+> NOTE: without an annotation to differentiate the operator behavior, the operator itself just creates the controller nodes in the pool and restarts the Kafka brokers to use them. ZooKeeper is not used anymore but no proper migration takes place (from ZooKeeper to KRaft metadata).
```

Comment:
My intention here was just reporting what's happening today.
Should we consider this as an issue today, so the need to address it which would take this out from the proposal?
Or maybe, if your current cluster has pools with only `broker` roles and no `controller`(s) pools we should detect the creation of a pool of controllers as an error if the migration is not set to be started. @scholzj any thoughts on this?

Reply:
Sorry, I don't follow the discussion. So I do not have any thoughts I'm afraid.

Reply:
I tried this scenario, and effectively the KRaft active controller takes over, and loads the existing metadata from ZK. No data loss.

```sh
2023-09-12 09:04:32,385 INFO [QuorumController id=5] Loaded ZK migration state of NONE (org.apache.kafka.controller.QuorumController) [quorum-controller-5-event-handler]
```

The problem is that the active controller is not in dual-write mode, so rollback to ZK controller is not possible anymore. We need to avoid this state for the migration, the contoller NodePool must be created only when the pre-migration annotation has been set, which means that we will include the migration configurations.

Reply:
Let's move the discussion out of the migration proposal. I opened the following issue on the operator https://github.com/strimzi/strimzi-kafka-operator/issues/9108